compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2713,6 +2714,42 @@ private void checkConstantAssignment(BLangExpression varRef, AnalyzerData data)
         }
     }
 
+    private boolean isFiniteWithNilMember(BType bType) {
+        if (!(bType.getKind() == TypeKind.FINITE)) {
+            return false;
+        }
+        BFiniteType finiteType = (BFiniteType) bType;
+        Set<BLangExpression> finiteMembers = finiteType.getValueSpace();
+        for (BLangExpression finiteMemberType : finiteMembers) {
+            if (finiteMemberType.getBType() != null && finiteMemberType.getBType().tag != TypeTags.NIL) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean isUnionWithSingleNilTypeMember(BType bType) {
+        if (!(bType.getKind() == TypeKind.UNION)) {
+            return false;
+        }
+        BUnionType unionType = (BUnionType) bType;
+        LinkedHashSet<BType> members = unionType.getMemberTypes();
+        LinkedHashSet<BType> originalMembers = unionType.getOriginalMemberTypes();
+        if (members.size() >= 1) {
+            for (BType memberType : members) {
+                if (memberType.getKind() != TypeKind.NIL) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        // To bypass `function f3() returns never|never` kind of scenarios, referencing error: #39797
```

Comment:
Let's just say this is a special case to handle return types like `never|never`, since they are not handled correctly at the moment and add the link to the issue (instead of just #39797).